src/internal/_set.js

```diff
@@ -0,0 +1,23 @@
+var _extend = require('./_extend');
+var _has = require('./_has');
+var _isArray = require('./_isArray');
+var isEmpty = require('../isEmpty');
+var keys = require('../keys');
+
+module.exports = function _set(update, original) {
+  if (_isArray(original) || _isArray(update)) {
+    return update;
+  }
+  var originalKey = keys(original);
+  var updateKeys = keys(update);
+  if (isEmpty(originalKey) || isEmpty(updateKeys)) {
+    return update;
+  }
+  var result = _extend({}, original);
```

Comment:
Just noting that `extend` should be doing much the same job as this function, less the recursive bit.   I didn't know if there would be a useful way to combine the looping inside `extend` with what's going on here.   But you might be able to chalk that up to random thoughts on a bus-ride.